# Example notebook

In [1]:
from noteLib import *
import datacube

dc= datacube.Datacube()

In [2]:
# dir(dc)

In [3]:
!datacube dataset search product ='l8_kline' |grep id |wc

    570    1140   23370


In [4]:
from datetime import datetime
date_range = (
        datetime(2014,6,13),
        datetime(2014,6,16))


In [5]:
x_lon_extent = (-119.44,-119.18)
y_lat_extent = (36.80,36.96)


In [6]:
product = 'l8_kline'
ds = dc.find_datasets(product=product,
             x = x_lon_extent, y =y_lat_extent,
             output_crs = 'epsg:32610', resolution = (-30,30), 
             time = date_range, measurements = ('red', 'green', 'blue','nir') )

In [7]:
myds = ds[0]

type(myds)

datacube.model.Dataset

In [8]:
print(len(ds))
print ("Number of Datasets")

1
Number of Datasets


In [9]:
print(myds.crs)

epsg:5072


In [10]:
for uri in myds.uris:
    print(uri)

s3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1.json


In [11]:
print(myds.extent)

Geometry({'type': 'Polygon', 'coordinates': [[(-2257500.0, 1747200.0), (-2257500.0, 2012700.0), (-1998000.0, 2012700.0), (-1998000.0, 1747200.0), (-2257500.0, 1747200.0)]]}, CRS('epsg:5072'))


In [12]:
print(myds.extent.type)

Polygon


In [13]:
print (type(myds.extent))

<class 'datacube.utils.geometry.Geometry'>


In [14]:
print (type(myds.extent.type))

<class 'str'>


In [15]:
print (myds.extent.boundingbox)

BoundingBox(left=-2257500.0, bottom=1747200.0, right=-1998000.0, top=2012700.0)


In [16]:
print (myds.extent.area)

68897250000.0


In [17]:
print (myds.extent.boundary)

Geometry({'type': 'LineString', 'coordinates': [(-2257500.0, 1747200.0), (-2257500.0, 2012700.0), (-1998000.0, 2012700.0), (-1998000.0, 1747200.0), (-2257500.0, 1747200.0)]}, CRS('epsg:5072'))


In [18]:
print (myds.extent.boundary.type)

LineString


In [19]:
print (myds.extent.coords)

None


# Should be something; I Think? --  Not None.

## Lets look at the stack of images - do a dc.load with the same parameters

In [20]:
product = 'l8_kline'
ds = dc.find_datasets(product=product,
             x = x_lon_extent, y =y_lat_extent,
             output_crs = 'epsg:32610', resolution = (-30,30), 
             time = date_range, measurements = ('red') )

In [21]:
print(len(ds))
print ("Number of Datasets")

1
Number of Datasets


In [22]:
myds = ds[0]

type(myds)

datacube.model.Dataset

In [23]:
print (myds.measurements)
print("---" * 35)
print("---" * 35)
print ("aka: band names and band file paths")


{'nir': {'path': 's3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_sr_band5.TIF', 'layer': 1}, 'swir1': {'path': 's3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_sr_band6.TIF', 'layer': 1}, 'therm': {'path': 's3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_bt_band10.TIF', 'layer': 1}, 'blue': {'path': 's3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_sr_band2.TIF', 'layer': 1}, 'swir2': {'path': 's3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_sr_band7.TIF', 'layer': 1}, 'green': {'path': 's3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_sr_band3.TIF', 'layer

In [24]:
print (myds.measurements['red'])

{'path': 's3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_sr_band4.TIF', 'layer': 1}


In [25]:
print (myds.measurements['red']['path'])

s3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_sr_band4.TIF


In [26]:
def return_path_row(file_path):
    # print(file_path)
    a = file_path.split('_')
    # print(a[0])
    print(a[2], a[9])
    return(a[2])

In [27]:
fp = myds.measurements['red']['path']

In [28]:
return_path_row(fp)


043034 20140615


'043034'

# Are they different path rows?

## Yes

In [29]:
for t in range(0,len(ds)):
    myds = ds[t]
    fp = myds.measurements['red']['path']
    return_path_row(fp)

043034 20140615


In [30]:
myds = ds[0]
print(myds.bounds)

BoundingBox(left=-2257500.0, bottom=1747200.0, right=-1998000.0, top=2012700.0)


In [31]:
ulX = myds.bounds.left
ulY = myds.bounds.top

lrX = myds.bounds.right
lrY = myds.bounds.bottom

In [32]:
m = aea_base_map(ulX,ulY,lrX,lrY)

In [33]:
aea_display_map(m,ulX,ulY,lrX,lrY)

In [34]:
date_range = (
        datetime(2014,6,13),
        datetime(2014,6,16))

x_lon_extent = (-120.0,-119.0)
y_lat_extent = (36.,36.96)


## Why 20140530 - another bug?

In [35]:
def display_bound(ds):
    
    myds = ds[0]

    ulX = myds.bounds.left
    ulY = myds.bounds.top

    lrX = myds.bounds.right
    lrY = myds.bounds.bottom
    m = aea_base_map(ulX,ulY,lrX,lrY)
    
    for i in range(0,len(ds)):
        myds = ds[i]

        ulX = myds.bounds.left
        ulY = myds.bounds.top

        lrX = myds.bounds.right
        lrY = myds.bounds.bottom
    
        aea_display_map(m,ulX,ulY,lrX,lrY)
        
    return(m)
        
        

In [36]:
date_range = (
        datetime(2014,5,13),
        datetime(2014,6,16))

x_lon_extent = (-119.44,-119.18)
y_lat_extent = (36.80,36.96)

product = 'l8_kline'
ds = dc.find_datasets(product=product,
             x = x_lon_extent, y =y_lat_extent,
             output_crs = 'epsg:32610', resolution = (-30,30), 
             time = date_range, measurements = ('red', 'green', 'blue','nir') )

In [37]:
m1 = display_bound(ds)
m1

In [38]:
for t in range(0,len(ds)):
    myds = ds[t]
    fp = myds.measurements['red']['path']
    return_path_row(fp)

042034 20140523
042034 20140608
042035 20140523
042035 20140608
043034 20140514
043034 20140530
043034 20140615


In [39]:
import prepare

In [40]:
ul_lon = x_lon_extent[0]
ul_lat = y_lat_extent[1]

lr_lon = x_lon_extent[1]
lr_lat = y_lat_extent[0]

In [41]:
from prepare import *
ulx,uly = ge_translate(ul_lat,ul_lon)

print(ulx,uly)

36.96 -119.44
-2046688.6812870034 1799970.259821642
-2046688.6812870034 1799970.259821642


In [42]:
lrx,lry = ge_translate(lr_lat,lr_lon)


36.8 -119.18
-2028770.5147079313 1777004.978578157


In [43]:
m = aea_base_map(ulx,uly,lrx,lry)

In [44]:
 aea_display_map(m,ulx,uly,lrx,lry)

In [45]:
aea_display_map(m,ulX,ulY,lrX,lrY, color='pink')

- UL Corner Lat dec: 38.52658
- UL Corner Long dec: -121.36464
- UR Corner Lat dec: 38.12722
- UR Corner Long dec: -119.2449

- LR Corner Lat dec: 36.4083
- LR Corner Long dec: -119.78328
- LL Corner Lat dec: 36.80638
- LL Corner Long dec: -121.85497

In [46]:
ee_foot_ulg_lat = 38.52658
ee_foot_ulg_lon = -121.36464

ee_foot_urg_lat = 38.12722
ee_foot_urg_lon = -119.2449

ee_foot_lrg_lat = 36.4083
ee_foot_lrg_lon = -119.78328

ee_foot_llg_lat = 36.80638
ee_foot_llg_lon = -121.85497



In [47]:

ulx,uly = ge_translate(ee_foot_ulg_lat,ee_foot_ulg_lon)


38.52658 -121.36464
-2164581.489788618 2012440.4580529404


In [48]:
lrx,lry = ge_translate(ee_foot_lrg_lat,ee_foot_lrg_lon)

36.4083 -119.78328
-2091345.9964462908 1747534.2183623333


In [49]:
aea_display_map(m,ulx,uly,lrx,lry, color='blue')

## yellow is the ee foot print

## Green is the extent from the metadata

##

In [50]:

geo_display_map(m, ee_foot_ulg_lat, ee_foot_ulg_lon,
               ee_foot_urg_lat, ee_foot_urg_lon,
               ee_foot_lrg_lat, ee_foot_lrg_lon,
               ee_foot_llg_lat, ee_foot_llg_lon)



In [51]:
! aws s3 ls lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1.json

2018-08-14 16:30:38      26793 LC08_L2SP_043034_20140615_20171210_01_A1.json


In [52]:
! aws s3 cp s3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1.json -

{
  "espa_metadata": {
    "@version": "2.0",
    "@xmlns": "http://espa.cr.usgs.gov/v2",
    "@xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
    "@xsi:schemaLocation": "http://espa.cr.usgs.gov/v2 http://espa.cr.usgs.gov/schema/espa_internal_metadata_v2_0.xsd",
    "global_metadata": {
      "data_provider": "USGS/EROS",
      "satellite": "LANDSAT_8",
      "instrument": "OLI_TIRS",
      "acquisition_date": "2014-06-15",
      "scene_center_time": "18:39:36.4523780Z",
      "level1_production_date": "2017-03-05T02:44:04Z",
      "solar_angles": {
        "@zenith": "22.750229",
        "@azimuth": "122.556496",
        "@units": "degrees"
      },
      "earth_sun_distance": "1.015752",
      "wrs": {
        "@system": "2",
        "@path": "43",
        "@row": "34"
      },
      "product_id": "LC08_L2SP_043034_20140615_20171210_01_A1",
      "lpgs_metadata_file": "LC08_L1TP_043034_20140615_20170305_01_A1_MTL.txt",
      "corner": [
        {
          "@location": "UL"

- BoundingBox(left=-2257500.0, bottom=1747200.0, right=-1998000.0, top=2012700.0)

In [53]:
! aws s3 ls lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_MTL.json

2018-08-14 16:30:38       9660 LC08_L2SP_043034_20140615_20171210_01_A1_MTL.json


In [54]:
! aws s3 cp s3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_MTL.json -

{
  "L1_METADATA_FILE": {
    "METADATA_FILE_INFO": {
      "ORIGIN": "Image courtesy of the U.S. Geological Survey",
      "REQUEST_ID": "0701703046872_00018",
      "LANDSAT_SCENE_ID": "LC80430342014166LGN01",
      "LANDSAT_PRODUCT_ID": "LC08_L1TP_043034_20140615_20170305_01_A1",
      "COLLECTION_NUMBER": "01",
      "FILE_DATE": "2017-03-05T02:44:04Z",
      "STATION_ID": "LGN",
      "PROCESSING_SOFTWARE_VERSION": "LPGS_2.7.0"
    },
    "PRODUCT_METADATA": {
      "DATA_TYPE": "L1TP",
      "COLLECTION_CATEGORY": "A1",
      "ELEVATION_SOURCE": "GLS2000",
      "OUTPUT_FORMAT": "GEOTIFF",
      "SPACECRAFT_ID": "LANDSAT_8",
      "SENSOR_ID": "OLI_TIRS",
      "WRS_PATH": "43",
      "WRS_ROW": "34",
      "NADIR_OFFNADIR": "NADIR",
      "TARGET_WRS_PATH": "43",
      "TARGET_WRS_ROW": "34",
      "DATE_ACQUIRED": "2014-06-15",
      "SCENE_CENTER_TIME": "18:39:36.4523780Z",
      "CORNER_UL_LAT_PRODUCT": "38.30581",
      "CORNER_UL_LON_PRODUCT": "-122.40018",
      "CORNER_UR

      "CORNER_UL_LAT_PRODUCT": "38.30581",
      "CORNER_UL_LON_PRODUCT": "-122.40018",
      "CORNER_UR_LAT_PRODUCT": "38.90669",
      "CORNER_UR_LON_PRODUCT": "-119.49499",
      
      "CORNER_LL_LAT_PRODUCT": "36.02438",
      "CORNER_LL_LON_PRODUCT": "-121.58491",
      "CORNER_LR_LAT_PRODUCT": "36.60744",
      "CORNER_LR_LON_PRODUCT": "-118.76185",
      
      
      
- ee_foot_ulg_lat = 38.52658
- ee_foot_ulg_lon = -121.36464

- ee_foot_urg_lat = 38.12722
- ee_foot_urg_lon = -119.2449

- ee_foot_lrg_lat = 36.4083
- ee_foot_lrg_lon = -119.78328

- ee_foot_llg_lat = 36.80638
- ee_foot_llg_lon = -121.85497


In [55]:
! aws s3 ls lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/

2018-08-14 16:30:38      21751 LC08_L1TP_043034_20140615_20170305_01_A1.xml
2018-08-14 16:30:38     117413 LC08_L1TP_043034_20140615_20170305_01_A1_ANG.txt
2018-08-14 16:30:37       8804 LC08_L1TP_043034_20140615_20170305_01_A1_MTL.txt
2018-08-14 16:30:38      26793 LC08_L2SP_043034_20140615_20171210_01_A1.json
2018-08-14 16:30:37     146417 LC08_L2SP_043034_20140615_20171210_01_A1_ANG.json
2018-08-14 16:30:38       9660 LC08_L2SP_043034_20140615_20171210_01_A1_MTL.json
2018-08-14 16:30:37      11776 LC08_L2SP_043034_20140615_20171210_01_A1_bt_band10.IMD
2018-08-14 16:30:37   43531802 LC08_L2SP_043034_20140615_20171210_01_A1_bt_band10.TIF
2018-08-14 16:30:38   10225802 LC08_L2SP_043034_20140615_20171210_01_A1_bt_band10.TIF.ovr
2018-08-14 16:30:37      11776 LC08_L2SP_043034_20140615_20171210_01_A1_bt_band11.IMD
2018-08-14 16:30:37   41975896 LC08_L2SP_043034_20140615_20171210_01_A1_bt_band11.TIF
2018-08-14 16:30:38    9908214 LC08_L2SP_043034_20140615_20171210_01_A1_bt_band11.TIF.ovr
2

In [56]:
! aws s3 ls lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L1TP_043034_20140615_20170305_01_A1_MTL.txt

2018-08-14 16:30:37       8804 LC08_L1TP_043034_20140615_20170305_01_A1_MTL.txt


In [57]:
! aws s3 cp s3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L1TP_043034_20140615_20170305_01_A1_MTL.txt -

GROUP = L1_METADATA_FILE
  GROUP = METADATA_FILE_INFO
    ORIGIN = "Image courtesy of the U.S. Geological Survey"
    REQUEST_ID = "0701703046872_00018"
    LANDSAT_SCENE_ID = "LC80430342014166LGN01"
    LANDSAT_PRODUCT_ID = "LC08_L1TP_043034_20140615_20170305_01_A1"
    COLLECTION_NUMBER = 01
    FILE_DATE = 2017-03-05T02:44:04Z
    STATION_ID = "LGN"
    PROCESSING_SOFTWARE_VERSION = "LPGS_2.7.0"
  END_GROUP = METADATA_FILE_INFO
  GROUP = PRODUCT_METADATA
    DATA_TYPE = "L1TP"
    COLLECTION_CATEGORY = "A1"
    ELEVATION_SOURCE = "GLS2000"
    OUTPUT_FORMAT = "GEOTIFF"
    SPACECRAFT_ID = "LANDSAT_8"
    SENSOR_ID = "OLI_TIRS"
    WRS_PATH = 43
    WRS_ROW = 34
    NADIR_OFFNADIR = "NADIR"
    TARGET_WRS_PATH = 43
    TARGET_WRS_ROW = 34
    DATE_ACQUIRED = 2014-06-15
    SCENE_CENTER_TIME = "18:39:36.4523780Z"
    CORNER_UL_LAT_PRODUCT = 38.30581
    CORNER_UL_LON_PRODUCT = -122.40018
    CORNER_UR_LAT_PRODUCT = 38.90669
    CORNER_UR_LON_PRODUCT = -119.49499
    CORNER_LL_LAT_PROD

In [ ]:
! aws s3 cp s3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L1TP_043034_20140615_20170305_01_A1_MTL.txt -

In [3]:
! ! aws s3 cp s3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_MTL.json -

{
  "L1_METADATA_FILE": {
    "METADATA_FILE_INFO": {
      "ORIGIN": "Image courtesy of the U.S. Geological Survey",
      "REQUEST_ID": "0701703046872_00018",
      "LANDSAT_SCENE_ID": "LC80430342014166LGN01",
      "LANDSAT_PRODUCT_ID": "LC08_L1TP_043034_20140615_20170305_01_A1",
      "COLLECTION_NUMBER": "01",
      "FILE_DATE": "2017-03-05T02:44:04Z",
      "STATION_ID": "LGN",
      "PROCESSING_SOFTWARE_VERSION": "LPGS_2.7.0"
    },
    "PRODUCT_METADATA": {
      "DATA_TYPE": "L1TP",
      "COLLECTION_CATEGORY": "A1",
      "ELEVATION_SOURCE": "GLS2000",
      "OUTPUT_FORMAT": "GEOTIFF",
      "SPACECRAFT_ID": "LANDSAT_8",
      "SENSOR_ID": "OLI_TIRS",
      "WRS_PATH": "43",
      "WRS_ROW": "34",
      "NADIR_OFFNADIR": "NADIR",
      "TARGET_WRS_PATH": "43",
      "TARGET_WRS_ROW": "34",
      "DATE_ACQUIRED": "2014-06-15",
      "SCENE_CENTER_TIME": "18:39:36.4523780Z",
      "CORNER_UL_LAT_PRODUCT": "38.30581",
      "CORNER_UL_LON_PRODUCT": "-122.40018",
      "CORNER_UR

In [4]:
! ! aws s3 cp s3://lsaa-staging-cog/L08/2014/043/034/LC08_L2SP_043034_20140615_20171210_01_A1/LC08_L2SP_043034_20140615_20171210_01_A1_ANG.json -

{
  "FILE_HEADER": {
    "LANDSAT_SCENE_ID": "LC80430342014166LGN01",
    "SPACECRAFT_ID": "LANDSAT_8",
    "NUMBER_OF_BANDS": "11",
    "BAND_LIST": [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11
    ]
  },
  "PROJECTION": {
    "ELLIPSOID_AXES": [
      6378137.0,
      6356752.3142
    ],
    "MAP_PROJECTION": "AEA",
    "PROJECTION_UNITS": "METERS",
    "DATUM": "WGS84",
    "ELLIPSOID": "WGS84",
    "PROJECTION_PARAMETERS": [
      0.0,
      0.0,
      29030000.0,
      45030000.0,
      -96000000.0,
      23000000.0,
      0.0,
      0.0,
      0.0,
      0.0,
      0.0,
      0.0,
      0.0,
      0.0,
      0.0
    ],
    "UL_CORNER": [
      -2257500.0,
      2012700.0
    ],
    "UR_CORNER": [
      -1998000.0,
      2012700.0
    ],
    "LL_CORNER": [
      -2257500.0,
      1747200.0
    ],
    "LR_CORNER": [
      -1998000.0,
      1747200.0
    ]
  },
  "EPHEMERIS": {
    "EPHEMERIS_EPOCH_YEAR": "2014",
    "EPHEMERI

      0.9206052,
      -0.4597918,
      -3.880972e-05,
      -4.916793e-06
    ],
    "BAND05_SCA01_LINE_DEN_COEF": [
      6.664834e-07,
      -2.955828e-06,
      -2.496006e-10,
      6.796324e-12
    ],
    "BAND05_SCA01_SAMP_NUM_COEF": [
      -1.25593,
      0.4838571,
      0.9113434,
      -0.004683211,
      -5.73548e-07
    ],
    "BAND05_SCA01_SAMP_DEN_COEF": [
      -4.044222e-06,
      -6.842146e-06,
      -1.544688e-06,
      -1.900066e-11
    ],
    "BAND05_SCA02_MEAN_HEIGHT": "1750.000",
    "BAND05_SCA02_MEAN_L1R_LINE_SAMP": [
      3763.889,
      257.214
    ],
    "BAND05_SCA02_MEAN_L1T_LINE_SAMP": [
      3111.462,
      2248.092
    ],
    "BAND05_SCA02_LINE_NUM_COEF": [
      0.02868083,
      0.9161848,
      -0.4676269,
      -0.0005439885,
      -3.499881e-06
    ],
    "BAND05_SCA02_LINE_DEN_COEF": [
      5.353781e-07,
      -1.995109e-06,
      4.101143e-10,
      6.688948e-12
    ],
    "BAND05_SCA02_SAMP_NUM_COEF": [
      -1.20979,
      0.4880087,
     